plugins/scale/wayfire/plugins/scale-transform.hpp

```diff
@@ -0,0 +1,181 @@
+#pragma once
+
+#include <wayfire/view-transform.hpp>
+#include <wayfire/nonstd/observer_ptr.h>
+#include <string>
+#include <list>
+#include <algorithm>
+
+namespace wf
+{
+/**
+ * Overlay function. It is called after the transformed view was already
+ * rendered.
+ *
+ * @param fb     The framebuffer to render to.
+ * @param damage Damaged region to render.
+ */
+using scale_view_overlay = std::function<void (const wf::framebuffer_t& fb,
+    const wf::region_t& damage)>;
+
+/**
+ * transformer used by scale -- it is an extension of the 2D transformer
+ * with the ability to add overlays
+ */
+class scale_transformer_t : public wf::view_2D
+{
+  public:
+    scale_transformer_t(wayfire_view view) : wf::view_2D(view)
+    {}
+    ~scale_transformer_t()
+    {}
+
+    struct padding
+    {
+        unsigned int top    = 0;
+        unsigned int left   = 0;
+        unsigned int bottom = 0;
+        unsigned int right  = 0;
+    };
+
+    uint32_t get_z_order() override
+    {
+        return wf::TRANSFORMER_HIGHLEVEL + 1;
+    }
+
+    /* render the transformed view and then add all overlays */
+    void render_with_damage(wf::texture_t src_tex, wlr_box src_box,
+        const wf::region_t& damage, const wf::framebuffer_t& target_fb) override
+    {
+        /* render the transformed view first */
+        view_transformer_t::render_with_damage(src_tex, src_box, damage, target_fb);
+
+        /* call all transformers */
+        for (auto& ol : overlays)
+        {
+            (*ol->ol)(target_fb, damage);
+        }
+    }
+
+    /**
+     * Add a new overlay that is rendered after this transform.
+     *
+     * @param ol      The callback function that does the rendering.
+     * @param z_order Relative order; overlays are called in order.
+     * @param pad     Any extra padding to use. Paddings are not cumulative,
+     *                i.e. if two overlays add a top padding, the larger
+     *                of the two values is used.
+     */
+    void add_overlay(std::shared_ptr<scale_view_overlay>& ol, int z_order,
+        const padding& pad = {0, 0, 0, 0})
+    {
+        auto it =
+            std::find_if(overlays.begin(), overlays.end(),
+                [z_order] (const auto& other)
+        {
+            return other->z_order >= z_order;
+        });
+
+        overlays.insert(it, std::make_unique<overlay>(ol, z_order, pad));
+        extend_padding(this->pad, pad);
+        view->damage();
+    }
+
+    void add_overlay(std::shared_ptr<scale_view_overlay>&& ol, int z_order,
+        const padding& pad = {0, 0, 0, 0})
+    {
+        add_overlay(ol, z_order, pad);
+    }
+
+    /* remove an existing overlay */
+    void rem_overlay(scale_view_overlay *ol)
```

Comment:
IMO this is inconsistent, if we pass shared ptrs for adding, we should use shared ptrs for removing too. (or unique_ptr + observer_ptrs, but maybe shared_ptrs are easier here)